lisa/tests/scheduler/load_tracking.py

```diff
@@ -971,7 +979,44 @@ def get_trace_cpu_util(self):
             phase_df = df_window(df, (start, end), method='pre', clip_window=True)
 
             for cpu in self.cpus:
-                util = phase_df[phase_df['cpu'] == cpu]['util']
+
+                cpu_phase_df = phase_df[phase_df['cpu'] == cpu].copy()
+
+                for task in tasks:
+                    if self.rtapp_profile[task].phases[phase-1].cpus.count(cpu):
+                        df_state_clipped = df_window(df_state[task], (start, end), method='pre', clip_window=True)
+                        cpu_phase_df = cpu_phase_df.join(df_state_clipped[[]], how='outer')
```

Comment:
extending the index of a series/dataframe with another index or series should be factored out in lisa.datautils to preserve readability, especially if it's relying on a hack like this join()

Reply:
OK, something like this?

```
diff --git a/lisa/datautils.py b/lisa/datautils.py
index a430bff53952..a47daf8794ee 100644
--- a/lisa/datautils.py
+++ b/lisa/datautils.py
@@ -401,6 +401,8 @@ from lisa.utils import TASK_COMM_MAX_LEN, groupby, deprecate
 
     return functools.reduce(merge, df_list)
 
+def df_extend_index(df1, df2):
+    return df1.join(df2[[]], how='outer')
 
 def _resolve_x(y, x):
     """
diff --git a/lisa/tests/scheduler/load_tracking.py b/lisa/tests/scheduler/load_tracking.py
index 8787f2f06ae2..c94d159a69b0 100644
--- a/lisa/tests/scheduler/load_tracking.py
+++ b/lisa/tests/scheduler/load_tracking.py
@@ -29,7 +29,7 @@ from lisa.tests.base import (
 )
 from lisa.target import Target
 from lisa.utils import ArtifactPath, groupby, ExekallTaggable
-from lisa.datautils import series_mean, df_window, df_filter_task_ids, series_refit_index, df_split_signals, df_refit_index, df_add_delta
+from lisa.datautils import series_mean, df_window, df_filter_task_ids, series_refit_index, df_split_signals, df_refit_index, df_add_delta, df_extend_index
 from lisa.wlgen.rta import RTA, Periodic, RTATask
 from lisa.trace import FtraceCollector, requires_events, may_use_events, MissingTraceEventError
 from lisa.analysis.load_tracking import LoadTrackingAnalysis
@@ -987,7 +987,7 @@ Time in seconds for util_avg to converge (i.e. ignored time)
                 for task, wlgen_task in self.rtapp_profile.items():
                     if cpu in wlgen_task.phases[phase-1].cpus:
                         df_state_clipped = df_window(df_state[task], (start, end), method='pre', clip_window=True)
-                        cpu_phase_df = cpu_phase_df.join(df_state_clipped[[]], how='outer')
+                        cpu_phase_df = df_extend_index(cpu_phase_df, df_state_clipped)
 
                 cpu_phase_df['cpu'].fillna(method='ffill', inplace=True)
                 cpu_phase_df = df_add_delta(cpu_phase_df)
```

Reply:
It would be good to have a function that works both on Series and DataFrame, you can take example on `_data_window()`/`df_window()`/`series_window()` in `lisa.datautils` on how it's done and documented:
* make an internal `_data_extend_index()` that works on both Series and DataFrame
* make a public `df_extend_index()` and `series_extend_index()` decorated with `DataFrameAccessor.register_accessor`/`SeriesAccessor.register_accessor`
* Document one of the public function and just refer to it in the docstring of the other one.

DataFrame and Series method are always separated in the public API to be able to add parameters to deal with pandas type-specific quirks if necessary (they can be very nasty and only show up in some conditions not obvious at the time of writing).

`_data_extend_index()` should operate on pandas Index types and the original object reindexed with the new merged index with `data.reindex()`:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.reindex.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reindex.html

Reply:
OK, will change this in the new version.